server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -163,20 +174,23 @@ private void addUnmappedFields(String[] indices, String field, Map<String, Field
             FieldCapabilities.Builder unmapped = new FieldCapabilities.Builder(field, "unmapped");
             typeMap.put("unmapped", unmapped);
             for (String index : unmappedIndices) {
-                unmapped.add(index, false, false, Collections.emptyMap());
+                unmapped.add(index, false, false, false, Collections.emptyMap());
             }
         }
     }
 
     private void innerMerge(Map<String, Map<String, FieldCapabilities.Builder>> responseMapBuilder,
-                                String indexName, Map<String, IndexFieldCapabilities> map) {
-        for (Map.Entry<String, IndexFieldCapabilities> entry : map.entrySet()) {
+                            FieldCapabilitiesIndexResponse response) {
+        for (Map.Entry<String, IndexFieldCapabilities> entry : response.get().entrySet()) {
             final String field = entry.getKey();
+            // best effort to detect metadata field coming from older nodes
+            final boolean isMetadataField = response.getOriginVersion().onOrAfter(Version.V_8_0_0) ?
+                entry.getValue().isMetadatafield() : metadataFieldPred.test(field);
```

Comment:
do we need this? could we just say that a field is a metadata field if it is in the node where it was taken from, based on its version?

Reply:
It's done to ensure that we don't break ML and QL during rolling upgrade. If there is a node that is on an older version, we try to mark the metadata fields coming from that node on a best effort basis. That allows to remove the static list in ML and QL since they can't be sure that the cluster was upgraded fully.